### ISSUE TITLE
Add safe list indexing operator: !?

### ISSUE THREAD

**Issue by phadej** [haskell/core-libraries-committee#110]
MR: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9511

```haskell
-- | A total variant of the list index function @('!!')@.
--
-- >>> [2,3,4] !? 1
-- Just 3
-- >>> [2,3,4] !? (-1)
-- Nothing
-- >>> []      !? 0
-- Nothing
--
(!?) :: [a] -> Int -> Maybe a
infixl 9 !?
```

This operator is reimplemented multiple times on Hackage:
- `safe`: https://hackage.haskell.org/package/safe/docs/Safe.html#v:atMay
- `extra`: https://hackage.haskell.org/package/extra/docs/Data-List-Extra.html#v:-33--63-
- `relude`: https://hackage.haskell.org/package/relude/docs/Relude-List.html#v:-33--33--63-
- `Agda`: https://hackage.haskell.org/package/Agda/docs/Agda-Utils-List.html#v:-33--33--33-
- Also `xmonad-contrib`, `ShellCheck`, `tidal`, ... https://hoogle.haskell.org/?hoogle=%5Ba%5D+-%3E+Int+-%3E+Maybe+a&scope=set%3Astackage

**Comment by treeowl** [comment]
I think I learned that operator name from `vector`, and it's now used in `containers` as well (did I do that? Memory fuzzy). Yes, I most definitely support this proposal.

**Comment by nomeata** [comment]
I'd suggest to put it first in the export list and phrase it's doc as if it is the now “normal” indexing operator, while `!!` is described as a variant of `!?` for use in situations where you are sure the indexing will succeed.

(But make sure the docs for `!!` are still useful on their own, e.g. in the haddocks of the Prelude )

**Comment by phadej** [comment]
@nomeata what you suggest concretely. Please suggest an edit for the exact wording on the MR. I'll let CLC to decide on that.

(Also my English is poor, the less prose I write, the less wrong it is).

**Comment by nomeata** [comment]
Absolutely right :-)

Since I had to touch parts not in your MR I opened a MR against your branch: https://gitlab.haskell.org/phadej/ghc/-/merge_requests/1

Merge if you like, or just take inspiration, as you like.

**Comment by phadej** [comment]
@Bodigrim there doesn't seem to be discussion, could you trigger a vote.

**Comment by nomeata** [comment]
What do think of my suggestions? Did you not want to merge before the committee deliberates on them?
(They are only docs changes, I can also submit them upstream after your change is merged, if that's preferable for some reason)

**Comment by phadej** [comment]
@nomeata TBH, I don't want to put much work until the whole issue is acknowledged by CLC to start with.

**Comment by mixphix** [comment]
Wholly in favour, this should have been the "main" operator from the beginning.

As an aside, though it would probably be too much breakage to be worth the hassle now, the fact that `(Data.List.!!)` has two exclamation points whereas everywhere else (`containers`, `vector`, `array`) the analogue is named `(!)` has always irked me.

**Comment by Bodigrim** [comment]
I'm fully in favor of the proposal, but I'd like to see an impact assessment. I imagine some packages may import `Data.List` unqualified, but define or use `(!?)` from elsewhere. After all `Data.List` is a part of Haskell Report, so even additive changes are worth additional scrutiny.

**Comment by Bodigrim** [comment]
And if we are to do an impact assessment anyway, maybe re-export `(!?)` from `Prelude` as well?

**Comment by phadej** [comment]
I kindly ask others to do impact assessment. I won't do it myself any time soon. (And probably not until https://github.com/Bodigrim/clc-stackage/issues/6 is resolved).

---

@Bodigrim 

> And if we are to do an impact assessment anyway, maybe re-export (!?) from Prelude as well?

That will definitely break *a lot* of stuff, e.g. `containers` and `vector` to begin with. E.g. https://hackage.haskell.org/package/vector-0.13.0.0/docs/src/Data.Vector.Fusion.Bundle.Monadic.html#%21%3F already hides alot of stuff from `Prelude` as `vector` defines name-clashing combinators, but because it's hiding and not importing only what is used, the `!?` will clash. Similarly `containers` `import Prelude hiding ...` in https://hackage.haskell.org/package/containers-0.6.6/docs/src/Data.Sequence.Internal.html#%21%3F

I would suggest these libraries to only import what's used, e.g. like in `vec`: https://github.com/phadej/vec/blob/0120d4ba43af22824d0bc4ab8c263daac358e5f0/vec/src/Data/Vec/Lazy.hs#L89-L93 to make them less impacted for `Prelude` additions, but that's up to their maintainers to do.

**Comment by phadej** [comment]
FWIW, this addition will break `extra`: https://hackage.haskell.org/package/extra-1.7.12/docs/src/Data.List.Extra.html
It imports `Data.List` unqualified (and re-exports it), but has `base >=4.9 && <5` bound. So technically **is violating PVP**, as any addition to `Data.List` leaks and isn't controlled by `extra` version. (An example is `Data.List.NonEmpty.singleton` addition: it leaks through `extra`'s `Data.List.NonEmpty.Extra`).

To be honest, that really sucks that we have to account for PVP-violating packages. The policy was put in place so the evolution of packages would be easier: addition only changes would break things. But as people don't care, that doesn't work.

It's very sad that in practice any change to `base` has to be treated as breaking change. That is unmotivating to say the least. I cannot force myself to write patch to the `extra` package, as it's an issue with how the package is written, not that the change in `base` is breaking one (per PVP).

I'll leave making this proposal to go through for more motivated people. I hoped it to be a small and quick improvement, but looks like it's not the case.

**Comment by emilypi** [comment]
> To be honest, that really sucks that we have to account for PVP-violating packages. 

Agreed. However, Neil is around and can presumably be persuaded to amend their bounds for this kind of quality of life improvement. I'd agree to this proposal.

**Comment by phadej** [comment]
> Agreed. However, Neil is around and can presumably be persuaded to amend their bound

`extra` is violating PVP as the API depends on the versions of dependencies, so either
- the bounds need to be quite strict, so the API of re-exported modules don't change in the range (this means quite tight lower bound as well as minor upper bounds).
- or better,  is to improve the code, so the imports are explicit (qualified or import list), as well as the re-exports.

I hope Neil view point has changed, but IIRC he was the one of first people to start using `base <5` bounds, re https://neilmitchell.blogspot.com/2014/11/upper-bounds-or-not.html

**Comment by emilypi** [comment]
@ndmitchell? ^

**Comment by ndmitchell** [comment]
I'm happy to add the necessary bounds if this gets added (both in a new release and in a revision) - I really wouldn't worry about `extra` to count as a point against this proposal.

**Comment by emilypi** [comment]
Thanks @ndmitchell!

**Comment by ekmett** [comment]
Anything of mine that breaks, I'll happily patch. Strong :+1: from me as a user!

**Comment by phadej** [comment]
I did quick build of clc-stackage and three packages failed:

```
cabal: Failed to build extra-1.7.10 (which is required by zenacy-html-2.0.4,
xlsx-tabular-0.2.2.1 and others). See the build log above for details.
Failed to build pcf-font-0.2.2.1. See the build log above for details.
Failed to build yaml-unscrambler-0.1.0.10 (which is required by
domain-0.1.1.3). See the build log above for details.
```

```
src/Data/List/Extra.hs:21:23: error:
    Ambiguous occurrence ‘!?’
    It could refer to
       either ‘Data.List.!?’,
              imported from ‘Data.List’ at src/Data/List/Extra.hs:41:1-16
              (and originally defined in ‘GHC.List’)
           or ‘Data.List.Extra.!?’,
              defined at src/Data/List/Extra.hs:150:4
   |
21 |     headDef, lastDef, (!?), notNull, list, unsnoc, cons, snoc,
   |                       ^^^^
```

```
src/Graphics/Text/PCF.hs:106:58: error:
    Ambiguous occurrence ‘!?’
    It could refer to
       either ‘Data.List.!?’,
              imported from ‘Data.List’ at src/Graphics/Text/PCF.hs:66:1-16
              (and originally defined in ‘GHC.List’)
           or ‘Data.Vector.!?’,
              imported from ‘Data.Vector’ at src/Graphics/Text/PCF.hs:72:21-24
    |
106 |         offset      <- fromIntegral <$> (bitmaps_offsets !? glyph_index)
    |                                                          ^^
```

```
library/YamlUnscrambler/Util/Vector.hs:8:61: error:
    Ambiguous occurrence ‘!?’
    It could refer to
       either ‘Data.Vector.!?’,
              imported from ‘Data.Vector’ at library/YamlUnscrambler/Util/Vector.hs:3:1-18
           or ‘Prelude.!?’,
              imported from ‘YamlUnscrambler.Prelude’ at library/YamlUnscrambler/Util/Vector.hs:4:1-57
              (and originally defined in ‘GHC.List’)
  |
8 |   getFirst (Prelude.foldMap (\k -> First (fmap (k,) (vector !? k))) keys)
  |                                                             ^^
```

All three are not adhering to PVP, specifically [not importing module properly](https://wiki.haskell.org/Import_modules_properly)

All three packages: `extra`, `pcf-font` and `yaml-unscrambler` import `Data.List` unqualified. (Also problem for #list There's only 40 or so packages depending on these three, rest of `clc-stackage` (~2800 packages) built fine.

I'll be happy to wait until https://github.com/haskell/core-libraries-committee/issues/22 is resolved, as that most likely will require (similar) patches to many more packages anyway.

**Comment by phadej** [comment]
@Bodigrim if it's not required to provide patches for these packages, then sure.

**Comment by Bodigrim** [comment]
It's trivial to write _some_ patches for the affected packages. I can do it myself if the proposal passes. 

Dear CLC members, let's vote on the proposal to extend `Data.List` with `(!?) :: [a] -> Int -> Maybe a` as detailed in https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9511/diffs. 
@tomjaguarpaw @chessai @cgibbard @emilypi @mixphix 

----

+1 from me.

**Comment by chessai** [comment]
+1

**Comment by emilypi** [comment]
+1

**Comment by tomjaguarpaw** [comment]
-1

----

I don't feel we shouldn't be encouraging users to index into lists, safely or unsafely. `!!` is probably too deeply embedded to ever be removable but we shouldn't perpetuate the issue by adding `!?`. I would rather we added a note that users who want to index into a list use a `Data.Sequence.Seq` instead.

**Comment by mixphix** [comment]
Indexing into a list is a `Natural` thing to want to do. Despite its linear complexity, it seems almost rude not to offer a way to get the `n`th element of `aList :: [a]` without reeling in a `containers` dependency.

+1

**Comment by tomjaguarpaw** [comment]
It's definitely a natural thing to want to do, which is why I think we should be discouraging it. The fewer footguns in `base` the better.

**Comment by Bodigrim** [comment]
4 votes out of 6 possible are enough to approve, thanks all.

(Not closing so that the open ticket is a remainder to prepare patches once we know whether this change lands in time for GHC 9.6 or not)

**Comment by emilypi** [comment]
@tomjaguarpaw while `!!` is definitely a footgun, `!?` hardly is, and the discussion about whether or not to include `!!` sailed before I was born. If the complaint is just one of complexity, I don't find that argument compelling - just provide a note in the comments regarding asymptotics in the haddocks. At the very least, it's good UX to provide `!?` and steer people towards that until they're ready to accept `Seq` as their lord and savior. Greeting various people from languages where indexing operations are business as usual with "but you're doing it all wrong just add the `containers` dependency to your build file and use `Seq`, the type of fast-indexable sequences of finite values!" is a serious complexity spike compared to the former option.

**Comment by tomjaguarpaw** [comment]
Naturally this is all moot, but for those who are interested ...

----

Yes, my objection is about complexity. No doubt `!?` is better than `!!`.

> just provide a note in the comments regarding asymptotics in the haddocks ... Greeting ... with "but you're doing it all wrong just add the containers dependency ... is a serious complexity spike compared to the former option.

That's a strong argument. However, given that there were three +1s before me I felt that adding a dissenting voice added more value here than just another +1.

**Comment by phadej** [comment]
`Seq` is not even always the right alternative type. In some cases it's some variant of array, (`Vector`, ...) or e.g. (skew) binary list. Proposing alternatives is *hard*, you have to know the use case.

**Comment by tomjaguarpaw** [comment]
Sure, my suggestion wasn't meant to be literal or exclusive. The suggestion doesn't have to be *only* `Seq`[^1]. There just has to be a big warning that `!!` (and now `!?`) is O(n) and a suggestion of alternative data structures, perhaps several.

O(n) indexing is a *big* footgun.  I disagree with Emily on that. Perhaps it's an even bigger than partiality of `!!` because at least with partiality of `!!` you know the bad thing is happening when you get an `error`. With O(n) indexing your program sometimes becomes slow and you have no idea why...

[^1]:  `Seq` is probably the best general purpose alternative to lists, and if `Vector` is the correct alternative in any particular case then indexing lists is an *even bigger* footgun in that case.

**Comment by phadej** [comment]
Add a big warning on `++` then too?

More seriously, `++` actually has a warning

```
WARNING: This function takes linear time in the number of elements of the first list.
```

so I added similar to `!!` and `!?`.

(but `length` doesn't, it has `O(n)` annotation instead. I'll leave it to `base` maintainers (you?) to harmonize the documentation).

**Comment by emilypi** [comment]
Thanks @phadej

**Comment by ekmett** [comment]
Oleg has been making a lot of hay about "PVP violating" packages here.

To be clear it **does not** violate the PVP to depend on a major version and use unqualified imports. It is merely deemed "slightly risky" by the PVP to do so and that you are on the "safe side" if you use qualified or explicit imports:

> To minimize breakage when new package versions are released, you can use dependencies that are insensitive to minor version changes (e.g. foo >= 1.2.1 && < 1.3). However, note that this approach is slightly risky: when a package exports more things than before, there is a chance that your code will fail to compile due to new name-clash errors. The risk from new name clashes may be small, but you are on the safe side if you [import identifiers explicitly or using qualification](https://wiki.haskell.org/Import_modules_properly).

The net effect is that if you do depend on a major version and breaking minor changes come down upstream then you are forced to do revisions. I personally find the revision updates to be less pain than the enduring soul-sucking effects of using qualified imports _all the time_ or thrashing explicit import lists to add and remove names as I code when I may touch hundreds of names in a module.

**Comment by phadej** [comment]
@ekmett, I disagree, your quote comes from section of PVP below the main points which are

> 4. Client specification. Hence A.B.C uniquely identifies the API. A client that wants to specify that they depend on a particular version of the API can specify a particular A.B.C and be sure of getting that API only. For example, build-depends: mypkg >= 2.1.1 && < 2.1.2.

> 5. Backwards compatible client specification. Often a package maintainer wants to add to an API without breaking backwards compatibility, and in that case they can follow the rules of point 2, and increase only C. A client MAY specify that they are [insensitive to additions to the API](https://wiki.haskell.org/Import_modules_properly) by allowing a range of C values, e.g. build-depends: base >= 2.1.1 && < 2.2.

**Comment by ekmett** [comment]
I think you read those two section quite differently than I.

(4) is permission for saying you don't care about changes across a known range of minor versions.

(5) is a 'MAY'. It is an explicit affordance to depend on a range of versions. It is not a MUST NOT about putting an upper bound on a major version alone in any circumstance other than such insensitivity. I realize that is splitting hairs, but a MAY in RFC 2119 lingo is simply not a prohibition of behavior that is not mentioned, but merely an explicit affordance of the behavior that is.

At the end of the day, I find the interpretation you apply to this wording far to onerous to comply with. Feel free to update the vocabulary of the PVP to make it even harder for me to write code in this language, but a written I do not agree with your interpretation.

**Comment by ekmett** [comment]
It wouldn't be "slightly risky" below if this was explicitly forbidden above, it would simply be against the policy. That is a _much_ stronger claim and not supported by the above wording.

**Comment by phadej** [comment]
@ekmett, my "violates" is indeed strong. The PVP says "MAY", so that requirement can be ignored, but if something goes wrong, it's the client's fault.

So, when you say

> Anything of mine that breaks, I'll happily patch. 

That is correct way to act (if it breaks due wild card imports), and I appreciate that. However, I don't think that additions to `base` (*not* `Prelude`) which would been prevented by explicit import list need an extra effort of patches from the community. People take risks, sometimes these risks realize.

Especially as wild card imports of `Data.List` have been warned about for other reasons for some time now, and also in `-Wall` lately.

**Comment by phadej** [comment]
> It wouldn't be "slightly risky"

I'd change it to just *risky*. These kind of name-clashing issues are not uncommon.

**Comment by Bodigrim** [comment]
Now we know that the change will be a part of `base-4.19` and GHC 9.8, I've raised PRs for all affected packages. Closing.
